File: demo-test/src/test/java/io/fries/demo/test/cucumber/steps/then/json/JsonStringAssert.java
package io.fries.demo.test.cucumber.steps.then.json;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import org.assertj.core.api.AbstractAssert;

import java.util.Set;
import java.util.stream.StreamSupport;

import static com.fasterxml.jackson.databind.SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS;
import static com.fasterxml.jackson.databind.SerializationFeature.WRITE_DATES_AS_TIMESTAMPS;
import static java.util.Comparator.comparing;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.fail;

public class JsonStringAssert extends AbstractAssert<JsonStringAssert, String> {

    private final ObjectMapper objectMapper;

    private JsonStringAssert(final String actual) {
        super(actual, JsonStringAssert.class);

        this.objectMapper = new ObjectMapper()
                .configure(ORDER_MAP_ENTRIES_BY_KEYS, true)
                .configure(WRITE_DATES_AS_TIMESTAMPS, false)
                .registerModule(new JavaTimeModule());
    }

    public static JsonStringAssert assertThatJsonString(final String actual) {
        return new JsonStringAssert(actual);
    }

    public JsonStringAssert isEqualToIgnoringFields(final String expected, final Set<String> fieldsToIgnore) {
        final var actualJsonNode = prepareJsonNode(actual, fieldsToIgnore);
        final var expectedJsonNode = prepareJsonNode(expected, fieldsToIgnore);

        assertThat(actualJsonNode.toPrettyString()).isEqualTo(expectedJsonNode.toPrettyString());
        return this;
    }

    private JsonNode prepareJsonNode(final String jsonString, final Set<String> fieldsToIgnore) {
        try {
            final var jsonNode = objectMapper.readTree(jsonString);
            return sortArraysOf(ignoreFieldsOf(jsonNode, fieldsToIgnore));
        } catch (final JsonProcessingException e) {
            return fail(e.getMessage(), e);
        }
    }

    private JsonNode ignoreFieldsOf(final JsonNode currentNode, final Set<String> fieldsToIgnore) {
        if (currentNode.isArray()) {
            currentNode.forEach(jsonNode -> ignoreFieldsOf(jsonNode, fieldsToIgnore));
        } else if (currentNode.isObject()) {
            currentNode.fields().forEachRemaining(field -> {
                if (fieldsToIgnore.contains(field.getKey())) {
                    ((ObjectNode) currentNode).replace(
                            field.getKey(),
                            objectMapper.convertValue("__value_ignored__", JsonNode.class)
                    );
                } else {
                    ignoreFieldsOf(field.getValue(), fieldsToIgnore);
                }
            });
        }

        return currentNode;
    }

    private JsonNode sortArraysOf(final JsonNode currentNode) {
        currentNode.fields().forEachRemaining(field -> sortArraysOf(field.getValue()));

        if (currentNode.isArray()) {
            final var arrayNode = (ArrayNode) currentNode;
            final var sortedArray = sort(arrayNode);
            arrayNode.removeAll().addAll(sortedArray);
        }

        return currentNode;
    }

    private ArrayNode sort(final ArrayNode arrayNode) {
        final var sortedNodes = StreamSupport.stream(arrayNode.spliterator(), false)
                .sorted(comparing(JsonNode::toString))
                .toList();

        return objectMapper
                .createArrayNode()
                .addAll(sortedNodes);
    }
}

File: api-test-proxy-wiremock/src/main/kotlin/io/fries/api/test/proxy/wiremock/validate/validation/json/JsonStringAssert.kt
package io.fries.api.test.proxy.wiremock.validate.validation.json

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS
import com.fasterxml.jackson.databind.SerializationFeature.WRITE_DATES_AS_TIMESTAMPS
import com.fasterxml.jackson.databind.node.ArrayNode
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import org.assertj.core.api.AbstractAssert
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.fail

class JsonStringAssert private constructor(actual: String?) :
    AbstractAssert<JsonStringAssert?, String?>(actual, JsonStringAssert::class.java) {

    private val objectMapper: ObjectMapper = ObjectMapper()
        .enable(ORDER_MAP_ENTRIES_BY_KEYS)
        .disable(WRITE_DATES_AS_TIMESTAMPS)
        .registerModule(JavaTimeModule())

    companion object {
        fun assertThatJsonString(actual: String?): JsonStringAssert = JsonStringAssert(actual)
    }

    fun isEqualToIgnoringFields(expected: String, fieldsToIgnore: Set<String>) = apply {
        val actualJsonNode = actual.prepareJsonNode(fieldsToIgnore)
        val expectedJsonNode = expected.prepareJsonNode(fieldsToIgnore)

        assertThat(actualJsonNode.toPrettyString()).isEqualTo(expectedJsonNode.toPrettyString())
    }

    private fun String?.prepareJsonNode(fieldsToIgnore: Set<String>): JsonNode = try {
        objectMapper
            .readTree(this)
            .ignoreFields(fieldsToIgnore)
            .sortArrays()
    } catch (e: JsonProcessingException) {
        fail(e.message, e)
    }

    private fun JsonNode.ignoreFields(fieldsToIgnore: Set<String>): JsonNode = apply {
        when (this) {
            is ArrayNode -> forEach { jsonNode -> jsonNode.ignoreFields(fieldsToIgnore) }
            is ObjectNode -> fields().forEachRemaining { (key, value) ->
                if (fieldsToIgnore.contains(key)) {
                    this.replace(key, objectMapper.convertValue("__value_ignored__", JsonNode::class.java))
                } else {
                    value.ignoreFields(fieldsToIgnore)
                }
            }
        }
    }

    private fun JsonNode.sortArrays(): JsonNode = apply {
        this.fields().forEachRemaining { (_, value) -> value.sortArrays() }
        if (isArray) {
            val arrayNode = this as ArrayNode
            val sortedArray = arrayNode.sort()
            arrayNode.removeAll().addAll(sortedArray)
        }
    }

    private fun ArrayNode.sort(): ArrayNode = objectMapper
        .createArrayNode()
        .addAll(sortedWith(Comparator.comparing { it.toString() }))
}